[ macOS ] webaudio/AudioBuffer/huge-buffer.html is a flaky time out.

jyavenard opened the pull request:
#### 114b7b67616c4f0bbd6ab6c329e1e15dc66b8eb0
<pre>
[ macOS ] webaudio/AudioBuffer/huge-buffer.html is a flaky time out.
<a href="https://bugs.webkit.org/show_bug.cgi?id=251710">https://bugs.webkit.org/show_bug.cgi?id=251710</a>
rdar://105028032

Reviewed by Jer Noble.

In r243819, the individual size of an audio channel was limited to 2^30
float samples, however an AudioBuffer can be made to have a high number
of channels.
The webaudio/AudioBuffer/huge-buffer.html test attempts to allocate an
AudioBuffer made of 32 channels with 2^30 float samples, this would require 2^37 bytes.
We limit the total size of an AudioBuffer to 2^32 float samples (2^34 bytes).

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/Modules/webaudio/AudioBuffer.cpp:
(WebCore::AudioBuffer::AudioBuffer):
* Source/WebCore/Modules/webaudio/AudioBuffer.h:

Canonical link: <a href="https://commits.webkit.org/261617@main">https://commits.webkit.org/261617@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b425bb1ad54e344bb226803ad21e1569ddc70c2f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112323 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21462 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/958 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4100 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120935 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116387 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22800 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/12526 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/4871 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118088 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/16940 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100125 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105370 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98897 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/661 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/45946 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13833 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/711 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/94344 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14526 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/10003 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19876 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52716 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8099 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16330 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/webaudio/AudioBuffer.cpp:
```diff
@@ -88,7 +89,8 @@ AudioBuffer::AudioBuffer(unsigned numberOfChannels, size_t length, float sampleR
     , m_isDetachable(preventDetaching == LegacyPreventDetaching::No)
 {
-    if (static_cast<uint64_t>(m_originalLength) > s_maxLength) {
+    auto totalLength = Checked<uint64_t, RecordOverflow>(m_originalLength) * numberOfChannels;
```
`CheckedUint64` alias exists for this.

jyavenard replied:
indeed, thanks

jyavenard committed:
[ macOS ] webaudio/AudioBuffer/huge-buffer.html is a flaky time out.
https://bugs.webkit.org/show_bug.cgi?id=251710
rdar://105028032

Reviewed by Jer Noble.

In r243819, the individual size of an audio channel was limited to 2^30
float samples, however an AudioBuffer can be made to have a high number
of channels.
The webaudio/AudioBuffer/huge-buffer.html test attempts to allocate an
AudioBuffer made of 32 channels with 2^30 float samples, this would require 2^37 bytes.
We limit the total size of an AudioBuffer to 2^32 float samples (2^34 bytes).

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/Modules/webaudio/AudioBuffer.cpp:
(WebCore::AudioBuffer::AudioBuffer):
* Source/WebCore/Modules/webaudio/AudioBuffer.h:

Canonical link: https://commits.webkit.org/261617@main

eric-carlson commented on Source/WebCore/Modules/webaudio/AudioBuffer.cpp:
```diff
@@ -117,13 +119,14 @@ AudioBuffer::AudioBuffer(AudioBus& bus)
     , m_originalLength(bus.length())
 {
-    if (static_cast<uint64_t>(m_originalLength) > s_maxLength) {
+    unsigned numberOfChannels = bus.numberOfChannels();
+    auto totalLength = CheckedUint64(m_originalLength) * numberOfChannels;
+    if (totalLength.hasOverflowed() || totalLength.value() > s_maxLength || static_cast<uint64_t>(m_originalLength) > s_maxChannelLength) {
```
It would be good to put this in a shared function rather than duplication this logic.
